### ISSUE TITLE
Inconsistency with .3ds files and .cia support

### ISSUE THREAD

**Issue by SamRohod** [SabreTools/NDecrypt#3]
Hi SabreTools
Thank you for creating this amazing easy-to-use all-in-one tool.

First, I use No-Intro database to verify that I have good files. When it comes to .nds and .dsi files, your tool always gives the correct hash values (md5) for both encrypted and decrypted files. But when it comes to .3ds files, it sometimes gives the correct md5 and other times not (for the same original file and for both encryption and decryption).   {I use NDecrypt-0.2.2_net48.7z}

I found the same problem for this tool (Batch CIA 3DS Decryptor):
https://gbatemp.net/threads/batch-cia-3ds-decryptor-a-simple-batch-file-to-decrypt-cia-3ds.512385/

But this tool (3DS + CIA Rom Script) seems consistent and always gives the same results as in No-Intro database:
https://anonfiles.com/Fb7fw8Xcob/3DS_CIA_Rom_Script_rar

Second, is it possible to add .cia support to your tool?
The two aforementioned tools do decrypt .cia files but the md5 is inconsistent and not matching No-Intro database.
Also neither of them encrypt .cia files

Thank you again

**Comment by mnadareski** [comment]
I don't check this repo often, so apologies that I missed this when you reported.

For your first part, if you can provide an example for something that didn't encrypt or decrypt properly (assuming you have a properly filled `keys.bin` file), that would help. Just a link to the Dat-o-Matic page would suffice. I had a user test every single 3DS file for both encrypt and decrypt and everything matched at the time.

For the second part, I can consider doing `.cia` support, but that's a little more complex. I don't mind eventually porting that sort of functionality. `.cia` files are structured differently than the `.3ds` files that are supported currently, so it's a non-trivial addition.

**Comment by SamRohod** [comment]
I'm Sorry for the mistake.
Apparently I had wrong keys.bin files.
I changed the title and the original issue post.

**Comment by XpertGame2020** [comment]
How to use the tool to decrypt .3ds files?

**Comment by mnadareski** [comment]
Not really, CIA support is still kinda low on my list given the other projects that I maintain. I appreciate the prod, though. Having the decrypt/encrypt working properly is quite a pain each time I've attempted it so far.

**Comment by Dimensional** [comment]
This is an old enhancement request. There's a python project called Ntool that supports converting CDN to CIA and back, and recently was able to work on making a reversible decryption routine for CIA creation. Turns out all one has to do is set the NoCrypto flag in the NCCH metadata without stripping out the other crypto flags, and altering minimal flags in the TMD, and it would still work.

**Comment by JohnTravolski** [comment]
> This is an old enhancement request. There's a python project called Ntool that supports converting CDN to CIA and back, and recently I was able to work on making a reversible decryption routine for CIA creation, meaning can switch between decrypted and encrypted. Turns out all one has to do is set the NoCrypto flag in the NCCH metadata without stripping out the other crypto flags, and altering minimal flags in the TMD with updated hashes, and it would still work. I've tested it myself. Porting to C# on is another story however.

This works great, thank you for making this!
